tests/types/resolving/test_optionals.py

```diff
@@ -16,6 +17,28 @@ def test_basic_optional():
     assert resolved == Optional[str]
 
 
+def test_optional_with_unset():
+    annotation = StrawberryAnnotation(Union[_Unset, Optional[str]])
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type is str
+
+    assert resolved == StrawberryOptional(of_type=str)
+    assert resolved == Optional[str]
+
+
+def test_optional_with_unset_as_union():
```

Comment:
Can we add a test like this for normal `Union`s (without `None`; e.g. `Union[SomeType, OtherType, _Unset]`)?

Reply:
Great idea!!

I was writing a test case for this but then I realized that if I specify a `Union[str, int, _Unset]`, it does not get stripped out (existing behavior) because the change in this PR is only occurring in the code path that hits optionals.

Optional types are detected by strawberry when it's a union that contains a `None`. We would need to add a `None` in this test case which doesn't make sense to add it to the union tests.

Instead, I've added another assertion into the `test_optionals`:

`test_optional_union_containing_a_real_union_and_unset`

This asserts that `Union[_Unset, None, Foo, Bar]` is resolved as a proper optional `Union[Foo, Bar]` without a default.

Reply:
Hmm that makes me wonder if this logic should instead be moved (or mirrored) to to `.create_union`.

Also, what happens if someone does `Optional[_Unset]` right now? I think that should probably throw an exception.

Reply:
Hmm that makes me wonder if this logic should instead be moved (or mirrored) to to `.create_union`.

Also, what happens if someone does `Optional[_Unset]` right now? I think that should probably throw an exception.

Reply:
IMO that is too much of an edge case to handle here since `_Unset` is already by design an internal type.

Reply:
Hmm, @patrick91 what do you think?

Reply:
sorry I missed this comment, I think it is fine to leave it! we can deal with this case when working on https://github.com/strawberry-graphql/strawberry/issues/1473